libraries/common-arm/src/error/mod.rs

```diff
@@ -0,0 +1,132 @@
+//! fn foo() -> Result<(), HydraError> {
+//!     // Do something that can throw an error
+//!     # Ok(())
+//! }
+//!
+//! let em = ErrorManager::new();
+//!
+//! em.handle(foo());
+//! ```
+//!
+//! ### RTIC
+//! [`ErrorManager`] is designed to be easily used with RTIC, and uses interior mutability to handle
+//! the errors. This makes it possible to share and access an instance without using locks:
+//! ```ignore
+//! #[task(shared = [&em])]
+//! fn foo(cx: foo::Context, m: Message) {
+//!     cx.shared.em.run(|| {
+//!         // ...
+//!     });
+//! }
+//! ```
+//!
+//! ## RTIC Task Spawning
+//! Spawning software tasks in RTIC can fail for various reasons. However, the `Result` returned
+//! by spawning a task is not standard, making it challenging to transform it into a
+//! [`HydraError`] while preserving useful info like the task name that failed to spawn. The macros
+//! [`spawn`](crate::spawn!), [`spawn_after`](crate::spawn_after!), and
+//! [`spawn_all`](crate::spawn_all!) are there to facilitate handling the errors that can be returned
+//! while spawning.
+//!
+//! [`spawn`](crate::spawn!) and [`spawn_after`](crate::spawn_after!) are macros that simply spawn
+//! the given task, converting the returned `Result` into a `Result<(), HydraError>`, adding in the
+//! process some useful info to the error like the task name.
+//!
+//! Most of the time, even if a single task fails to spawn, we should still try spawning the
+//! remaining tasks. The [`spawn_all`](crate::spawn_all!) macro will try to spawn all the tasks
+//! given to it. If all tasks were spawned successfully, it will return `Ok(())`. However, if a task
+//! could not spawn, it will return it's associated error. Note that if multiple tasks could not spawn,
+//! only the last error will be returned.
+//!
+//! Together, those macros would be used like this:
+//! ```ignore
+//! spawn_all!(
+//!     spawn!(send_message, message);
+//!     spawn_after!(sensor_send, 2.secs());
+//! )?;
+//! ```
+//!
+
+pub mod error_manager;
+pub mod hydra_error;
+
+pub use error_manager::ErrorManager;
+pub use hydra_error::HydraError;
+
+/// Calls `spawn` on a RTIC's task. Transforms the returned `Result` into a `Result<_, HydraError>`.
+#[macro_export]
+macro_rules! spawn {
+    ($task:ident$(,)? $($arg:expr),*) => {
+        $crate::SpawnError::spawn_error($task::spawn($($arg),*), stringify!($task))
+    };
+}
+
+/// Calls `spawn_after` on a RTIC's task. Transforms the returned `Result` into a `Result<_, HydraError>`.
+#[macro_export]
+macro_rules! spawn_after {
+    ($task:ident$(,)? $($arg:expr),*) => {
+        $crate::SpawnError::spawn_error($task::spawn_after($($arg),*), stringify!($task))
+    };
+}
+
+/// Runs all of the contained spawn statements. Returns `Ok(()` if all spawns succeeded. If a spawn
+/// failed, returns `Err(e)` containing the relevant error. If multiple spawns failed, only one
+/// error is returned.
+#[macro_export]
+macro_rules! spawn_all {
```

Comment:
How necessary did you find these macros to be? I'm especially looking at the spawn_all macro, which seems like it isn't totally needed. I'm hesitant to add macros willy-nilly because they're much harder to audit and modify, especially when starting out.

Reply:
`spawn` macros I think are necessary to make it easy to get a useful error out of them. 

For `spawn_all`, yeah, I don't know. I realized handling errors from spawning a task is a bit different from other errors. I think the usual pattern will be to spawn all tasks first, then check if any tasks returned an error. To make that pattern easier to use, I created the macro. However, I'm now realizing this is similar to a `finally` block, which actually this is not quite succeeding at doing. Do you have a better idea?

And to clarify, I'm trying to prevent this (because a failure in send_message will prevent `blink` from starting):
```rust
spawn!("send_message", message)?;
spawn!("blink")?;
```

In favour of something like this:
```rust
spawn_all!(
spawn!("send_message", message);
spawn!("blink");
)?;
```
Or maybe even something like this (pseudo-code):
```rust
try {
  // Complicated logic
} finally {
   spawn!("blink"); // We have a task that we should always try to start, say because it is on a timer
}
```

Now thinking more about it, maybe we can remove the `spawn_all!` macro for now, and only add it once we actually start encountering this problem.

Reply:
I think I agree about the `spawn` macros being a good idea and sure, let's hold off on the `spawn_all!` macro until we really need it.